### PR TITLE
SMLreader Binding

### PR DESCRIPTION
**Pull request by gilhmauy**
Initial contribution of SmlreaderBinding

### CODE REVIEW

**Comment by gilhmauy** [comment]
Documentation of the SmlReader binding bundle

## Introduction

SML is a communication protocol for transmitting meter data for so called smart meters. SML is similar to XML with regard to its ability to nest information tags within other tags. But the standard way of coding SML packets is in a more efficient binary format that is similar but not equal to BER encoding. Also the specification comes in a form that is similar to ASN.1 but not completely standards compliant.

To decode SML messages, the binding takes advantage of jSML library by [Fraunhofer Institute](http://www.openmuc.org/index.php?id=63).

For installation of the binding, please see Wiki page [Bindings](https://github.com/openhab/openhab/wiki/Bindings).

## SmlReader Binding Configuration
### openhab.cfg
The openhab.cfg file (in the folder '${openhab_home}/configurations') contain following config parameters  to configure the binding.
The configuration allows the definition of multiple SML devices.

    #########################  SmlReader Binding ####################
    # Perform a module status query every x seconds (optional, defaults to 600 (10 minutes)).
    smlreader:refresh=60

      # the serial port to use for connecting to the metering device e.g. "COM1" for Windows and "/dev/ttyS0" or
    # "/dev/ttyUSB0" for Linux
    smlreader:<smldevice-id1>.serialPort=COM1

## Generic Item Binding Configuration

### Item Binding Configuration

In order to bind an item to the data received from the meter device , you need to provide item configuration settings. The easiest way to do so is to add some binding information in your item file (in the folder configurations/items`). The syntax of the binding configuration strings accepted is the following:
The syntax of an item configuration is shown in the following line in general:

    smlreader="device=<smldevice-id1>,obis=<obis-id>"

Where "<smldevice-id1>" matches one of the ids defined in your openhab.cfg file.

SmlReaderBinding also supports the [openHAB transformation Service](https://github.com/openhab/openhab/wiki/Transformations) as well. To transform a SML value returned by the configured device just add transformation information to your item-configuration:

    smlreader="device=<smldevice-id1>,obis=<obis-id>,transform=<tranformation-service>(<transformation-file.js>)"

Where "<transformation-service>" is your preferred Transformation service and "<transformation-file>" is the file, where your transorfmation logic takes place. Transformation files have to be placed in '${openhab_home}/configurations/transform'.
The following transformation-services are supported in general:

* JS
* REGEX
* XSLT

The binding will dump all available OBIS codes with their corresponding value on the first refresh or on configuration changes in the osgi console. For further information about OBIS codes read [Wiki page](http://de.wikipedia.org/wiki/OBIS-Kennzahlen).

### Examples

Below is an example of represent the current energy counter tarif values as numeric items.

    /* SmlReader data */
    Number Obis_180 "Bezug total [%d kWh]" (gEnergy) { smlreader="device=edl300,obis=1-0:1.8.0,transform=JS(smlreader_kilo.js)" }
    Number Obis_280 "Einsp. total [%d kWh]" (gEnergy) { smlreader="device=edl300,obis=1-0:2.8.0,transfrom=JS(smlreader_kilo.js)" }
    Number Obis_1507 "Aktuell [%d W]" (gEnergy) { smlreader="device=edl300,obis=1-0:15.0.7" }

### Known limitations/issues

* SmlReaderBinding isn't able to read SML from devices which are working in "pull" mode, e.g. Froetec ZG22.
* Octet encoding for OBIS Codes
  - '129-129:199.130.5'
  - '1-0:0.0.9'
doesn't work properly.

Any help/contribution is appreciated!

### Tested Hardware

The binding has been successfully tested with below hardware configuration:
* EMH EDL300 meter connected the [IR-Reader USB](http://wiki.volkszaehler.org/hardware/controllers/ir-schreib-lesekopf-usb-ausgang) from open hardware project in [volkszaehler](http://volkszaehler.org/) to a Windows Server 2012 R2 Essentials

**Comment by watou** [comment]
Hi @gilhmauy,
It looks like there is a potential tester for your binding [here](https://community.openhab.org/t/integrate-power-meter-data-into-openhab/3351).  Could you provide him a link to a JAR for testing?

**Comment by teichsta** [comment]
Iit seems there were some issues during testing @gilhmauy. Anything that should be integrated into this PR?

**Comment by gilhmauy** [comment]
Hi @teichsta 
i've fixed the reported issue locally already but i'm struggling with git once again to commit the change. no idea actually...

File: bundles/binding/pom.xml [hunk 1]
```diff
@@ -158,5 +158,6 @@
     <module>org.openhab.binding.lightwaverf</module>
 	<module>org.openhab.binding.souliss</module>
     <module>org.openhab.binding.sagercaster</module>
+    <module>org.openhab.binding.smlreader</module>
```
Comment (zobelhelas) [inline bundles/binding/pom.xml]: this hunk does't apply against 1.8.2 any more.

File: distribution/openhabhome/configurations/openhab_default.cfg [hunk 2]
```diff
@@ -1982,3 +1982,12 @@ tcp:refreshinterval=250
 # IP address of a Panasonic TV instance
 # panasonictv:<id>=
 
+#########################  SmlReader Binding ####################
```
Comment (zobelhelas) [inline distribution/openhabhome/configurations/openhab_default.cfg]: this hunk does't apply against 1.8.2 any more.

File: distribution/pom.xml [hunk 3]
```diff
@@ -1293,6 +1293,12 @@
             <version>${project.version}</version>
             <type>jar</type>
         </dependency>
+        <dependency>
```
Comment (zobelhelas) [inline distribution/pom.xml]: this hunk does't apply against 1.8.2 any more.

**Comment by 9037568** [comment]
@gilhmauy , are you planning to finish the changes for this PR?

**Comment by kaikreuzer** [comment]
I assume this can be closed since we have https://github.com/openhab/openhab2-addons/pull/2167 as a follow-up.

**Pull request closed without merging**